### ISSUE TITLE
Register a cluster to Karmada with Pull mode failed

### ISSUE THREAD

**Issue by soulseen** [karmada-io/karmada#3280]
**What happened**:
Register a cluster to Karmada with Pull mode failed, error:
```
./karmadactl register karmada-apiserver.karmada-system.svc.cluster.local:5443 --token 5zmvck.151hyrw55gebxq11 --discovery-token-ca-cert-hash sha256:f2596eb0ff5c8acac0d292837889dc7bbcb5adc947d97bc3ec189f4ebdb0cf6a
[preflight] Running pre-flight checks
[prefligt] All pre-flight checks were passed
[karmada-agent-start] Waiting to perform the TLS Bootstrap



error: couldn't validate the identity of the API Server: configmaps "cluster-info" is forbidden: User "system:anonymous" cannot get resource "configmaps" in API group "" in the namespace "kube-public"
```
**What you expected to happen**:
register success
**How to reproduce it (as minimally and precisely as possible)**:
1. install karmada control plane by helm: https://github.com/karmada-io/karmada/tree/master/charts/karmada
2. Register a cluster to Karmada with Pull mode with command:
```
./karmadactl register karmada-apiserver.karmada-system.svc.cluster.local:5443 --token 5zmvck.151hyrw55gebxq11 --discovery-token-ca-cert-hash sha256:f2596eb0ff5c8acac0d292837889dc7bbcb5adc947d97bc3ec189f4ebdb0cf6a
```
**Anything else we need to know?**:

**Environment**:
- Karmada version: 
```
v1.5.0
```
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
```
karmadactl version: version.Info{GitVersion:"v1.5.0", GitCommit:"51fbbb055d47331e83a6ca2e90b8a5312d6a7986", GitTreeState:"clean", BuildDate:"2023-03-09T03:28:23Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
```
- Others:

**Comment by RainbowMango** [comment]
I remember `karmadactl register` is limited to registering a cluster to the control plane installed by `karmadaclt init`. Am I right? @lonelyCZ 

@Poor12 Do you have any suggestion to register a pull-mode cluster to the control plane installed by helm chart?

**Comment by Poor12** [comment]
Yes, it only support control plane installed by `karmadactl init` now according to https://karmada.io/docs/userguide/clustermanager/cluster-registration#register-cluster-by-cli-1. I remember it's because the certs are organized differently. cc expert @lonelyCZ .

**Comment by lonelyCZ** [comment]
> Ok, thanks, but I think the different installation methods should keep the consistent results.

Yes, I also have a plan to unify them in next version recently. Thanks for your suggestions @soulseen 

/assign

**Comment by RainbowMango** [comment]
@soulseen Is it a blocking issue for you?

**Comment by soulseen** [comment]
after I switched to using `karmadactl init`, it succeed. and registering a member cluster with pull mode, the log shows like:
```
[preflight] Running pre-flight checks
[prefligt] All pre-flight checks were passed
[karmada-agent-start] Waiting to perform the TLS Bootstrap
[karmada-agent-start] Waiting to construct karmada-agent kubeconfig
[karmada-agent-start] Waiting the necessary secret and RBAC
[karmada-agent-start] Waiting karmada-agent Deployment
W0315 17:52:25.156444 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:26.161413 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:27.160183 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:28.160190 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:29.159406 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:30.160753 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:31.160356 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:32.160247 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:33.159958 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:34.160028 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:35.159607 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:36.160598 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:37.159985 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:38.159790 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:39.160062 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:40.159833 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
W0315 17:52:41.159676 1396008 check.go:52] Pod: karmada-agent-bc88677f5-f4tvh not ready. status: ContainerCreating
I0315 17:52:42.158892 1396008 check.go:49] pod: karmada-agent-bc88677f5-f4tvh is ready. status: Running

cluster(k8s-danlu-vm-dev-new) is joined successfully
```
it seems like join success(k8s-danlu-vm-dev-new), but can not find the member cluster info in the control plane:
```
$ kubectl --kubeconfig /home/karmada-data/karmada-apiserver.config get clusters
NAME      VERSION   MODE   READY   AGE
qa-live   v1.18.3   Push   True    33m
```
@RainbowMango can you help to take a look? thanks

**Comment by RainbowMango** [comment]
How long did you check the cluster after registering it? The `register` returns immediately after launching the `karmada-agent`, but  it might take some time (not very long, 1 min should be enough ) for `karmada-agent` to register to the control plane.

Could you please share the logs of `karmada-agent` deployed by `register` command?

**Comment by soulseen** [comment]
`karmada-agent` log:
```
$ k -n karmada-system logs -f karmada-agent-bc88677f5-lglkq
I0316 01:48:51.695291       1 agent.go:115] karmada-agent version: version.Info{GitVersion:"v1.5.0", GitCommit:"51fbbb055d47331e83a6ca2e90b8a5312d6a7986", GitTreeState:"clean", BuildDate:"2023-02-28T12:02:49Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
E0316 01:49:21.709657       1 run.go:74] "command failed" err="Get \"https://192.168.12.23:35642/apis/cluster.karmada.io/v1alpha1/clusters\": dial tcp 192.168.12.23:35642: i/o timeout"
```
but use `telnet` to check TCP,  it works well:
```
$ telnet 192.168.12.23 35642
Trying 192.168.12.23...
Connected to 192.168.12.23.
Escape character is '^]'.
```

**Comment by soulseen** [comment]
it joins success after using `hostNetwork`. Now everything looks good.

**Comment by lonelyCZ** [comment]
> it joins success after using `hostNetwork` in `karmada-agent`. Now everything looks good.

Congratulation!

Is `hostNetwork` IP/Port of `karmada-apiserver` you're using?

**Comment by RainbowMango** [comment]
> it joins success after using hostNetwork in karmada-agent.

Could you please share your whole `karmadactl register` command?  I'm curious about the way how you use `hostNetwork`.

**Comment by soulseen** [comment]
> > it joins success after using `hostNetwork` in `karmada-agent`. Now everything looks good.
> 
> Congratulation!
> 
> Is `hostNetwork` IP/Port of `karmada-apiserver` you're using?

no, but I used nodePort svc of `karmada-apiserver`, and only one node in the member cluster can access the control plane node.

**Comment by soulseen** [comment]
> > it joins success after using hostNetwork in karmada-agent.
> 
> Could you please share your whole `karmadactl register` command? I'm curious about the way how you use `hostNetwork`.

I directly modified the deployment of `karmada-agent`.

**Comment by lonelyCZ** [comment]
> only one node in the member cluster can access the control plane node.

Ok, you need to deploy the agent in this node. 
> I did not use the command to set hostNetwork, and directly modified the deployment of karmada-agent.

I guess it is because your network is different, but why it is useful after `karmada-agent` is modified to `hostNetwork`? Could it be that the pod was scheduled to a new node that can access `karmada-apiserver` after you modified the deployment?

Generally, `karmada-agent` is just needed to access `karmada-apiserver` as client.

**Comment by soulseen** [comment]
@lonelyCZ Yes, and I missed the point that I not only changed the `hostnetwork` but also the `nodeselector` at the same time. 
Maybe I only need to update `nodeselector`.

**Comment by RainbowMango** [comment]
@lonelyCZ Are there any other tasks that need to be done for this feature?

**Comment by RainbowMango** [comment]
/close

Seems this has been solved by #3440, #3552, #3536.
